charon/src/ast/types.rs

```diff
@@ -747,12 +747,16 @@ pub enum TyKind {
     ///
     /// TODO: we don't translate this properly yet.
     DynTrait(ExistentialPredicate),
-    /// Arrow type, used for function pointers and reused for the unique type associated with each
-    /// function item.
+    /// Function pointer type. This is a literal pointer to a region of memory that
+    /// contains a callable function.
     /// This is a function signature with limited generics: it only supports lifetime generics, not
-    /// other kinds of
-    /// generics.
+    /// other kinds of generics.
     Arrow(RegionBinder<(Vec<Ty>, Ty)>),
+    /// The unique type associated with each function item. Each function item is given
+    /// a unique generic type that takes as input the function's early-bound generics. This type
+    /// is not generally nameable in Rust; it's a ZST (there's a unique value), and a value of that type
+    /// can be cast to a function pointer or passed to functions that expect `FnOnce`/`FnMut`/`Fn` parameters.
+    FnDef(RegionBinder<FunDeclRef>),
```

Comment:
```suggestion
    /// There's a binder here because charon function items bind both early and late-bound variables;
    /// that binder binds the late-bound lifetimes (those that could appear in a function pointer type
    /// like `for<'a> fn(&'a u32)`).
    FnDef(RegionBinder<FunDeclRef>),
```